test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,390 @@
+
+	var newIP string
+	for ip, mapped := range domainMapping {
+		if ip != fqdnIp && mapped == false {
+			newIP = ip
+		}
+	}
+
+	t.Logf("New IP to update to DNS %v", newIP)
+
+	// Curl the old ip and it should be a success.
+	err = curlFqdn(toolBoxPodName, toolboxContainerName, fqdnIp, true)
+	require.NoError(t, err, "failed to curl FQDN from antrea toolbox on initial run : %v", err)
+
+	// Update the custom DNS configMap
+	UpdatedCustomDNSconfig := createDnsConfig(newIP)
+
+	customDnsConfigMapObject.Data = UpdatedCustomDNSconfig
+	err = data.UpdateConfigMap(customDnsConfigMapObject)
+	require.NoError(t, err, "failed to update configmap with new IP : %v", err)
+	t.Logf("successfully updated dns configMap with new IP : %+v", newIP)
+
+	updateDnsPodAnnotations(t, data)
+
+	for {
+		err = curlFqdn(toolBoxPodName, toolboxContainerName, fqdnIp, true)
+		require.NoError(t, err, "curl to ip failed with error : %v", err)
+
+		// Wait for 1 second before retrying
+		time.Sleep(1 * time.Second)
+	}
+
+}
+
+func tearDown(t *testing.T, data *TestData, builder *utils.ClusterNetworkPolicySpecBuilder) {
+	// cleanup test resources
+	teardownTest(t, data)
+	failOnError(k8sUtils.DeleteACNP(builder.Name), t)
+}
+
+func getCustomDnsServerPodLabel() map[string]string {
+	return map[string]string{"app": "custom-dns"}
+}
+
+func createCustomDnsService(data *TestData) (*corev1.Service, error) {
+	return testData.CreateUDPService(customDnsServiceName, data.testNamespace, customDnsPort, customDnsTargetPort,
+		getCustomDnsServerPodLabel(), false, false, corev1.ServiceTypeClusterIP, getIPFamily())
+}
+
+// CreateUDPService creates a service with a UDP port and targetPort.
+func (data *TestData) CreateUDPService(serviceName, namespace string, port, targetPort int32, selector map[string]string,
+	affinity, nodeLocalExternal bool,
+	serviceType corev1.ServiceType, ipFamily *corev1.IPFamily) (*corev1.Service, error) {
+	annotation := make(map[string]string)
+	return data.CreateServiceWithAnnotations(serviceName, namespace, port, targetPort, corev1.ProtocolUDP, selector,
+		affinity, nodeLocalExternal, serviceType, ipFamily, annotation)
+}
+
+func getCustomDnsServiceIP(t *testing.T, data *TestData, customDnsService *corev1.Service) string {
+	// Usually the IP gets assigned quickly to a service , but maybe it's a safe idea to use Get() for getting the IP.
+	customCoreDnsServiceObject, err := data.clientset.CoreV1().Services(data.testNamespace).Get(context.Background(),
+		customDnsService.Name, metav1.GetOptions{})
+	require.NoError(t, err, "Error when getting custom DNS service object : %v", err)
+
+	if customCoreDnsServiceObject.Spec.ClusterIP == "" {
+		require.Fail(t, "ClusterIP is empty for the custom DNS service")
+	}
+
+	t.Logf("ClusterIP of the dns service: %s\n", customCoreDnsServiceObject.Spec.ClusterIP)
+	return customCoreDnsServiceObject.Spec.ClusterIP
+}
+
+func getIPFamily() *corev1.IPFamily {
+	ipFamily := corev1.IPv4Protocol
+	return &ipFamily
+}
```

Comment:
IMO, all of these functions are not necessary. The code could be much less verbose.
For example, can you check if `Spec.ClusterIP` is already populated in the return value of `CreateServiceWithAnnotations`?